crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -114,30 +425,123 @@ impl<S: States> Deref for State<S> {
     }
 }
 
+#[derive(Resource, Default, Debug)]
+#[cfg_attr(
+    feature = "bevy_reflect",
+    derive(bevy_reflect::Reflect),
+    reflect(Resource, Default)
+)]
+struct PreviousState<S: States>(S);
+
+impl<S: States> PreviousState<S> {
+    /// Get the current state.
+    pub fn get(&self) -> &S {
+        &self.0
+    }
+}
+
+impl<S: States + PartialEq> PartialEq<S> for PreviousState<S> {
+    fn eq(&self, other: &S) -> bool {
+        self.get() == other
+    }
+}
+
+impl<S: States> Deref for PreviousState<S> {
+    type Target = S;
+
+    fn deref(&self) -> &Self::Target {
+        self.get()
+    }
+}
+
 /// The next state of [`State<S>`].
 ///
 /// To queue a transition, just set the contained value to `Some(next_state)`.
 /// Note that these transitions can be overridden by other systems:
 /// only the actual value of this resource at the time of [`apply_state_transition`] matters.
-#[derive(Resource, Default, Debug)]
+#[derive(Resource, Default, Clone)]
 #[cfg_attr(
     feature = "bevy_reflect",
     derive(bevy_reflect::Reflect),
     reflect(Resource, Default)
 )]
-pub struct NextState<S: States>(pub Option<S>);
+pub enum NextState<S: States> {
+    /// Do not change the state.
+    #[default]
+    Keep,
+    /// Change the state to a specific, pre-determined value
+    Value(S),
+    /// Change the state to a value determined by the given closure
+    Setter(Arc<dyn Fn(S) -> S + Sync + Send>),
```

Comment:
Ditto on the Arc here: I think you should be able to convince Rust that this function itself must be Send + Sync.

Reply:
I was able to move to Box by removing the Clone from `NextState`. However, unlike with the StateMatcherFunction which was replaced with just an `fn` pointer, I don't think we can remove the ability to use closures.

The StateMatcherFunction's (and all state matcher stuff, really) is meant to be defined during initialization, and so should be static - which makes it easy to use a pointer for it. However, I can imagine a situation where I rely on a contextual variable within a "NextState" closure, for example:

```rust
fn start_game(mut event: EventReader<StartGameEvent>, mut next: ResMut<NextState<AppState>) {
  let Some(StartGameEvent { player_id }) = event.get_single() else {return; };
  next.setter(move |_| AppState::InGame { player_id });
}
```